### ISSUE TITLE
setTimeoutInfinite in brpop

### ISSUE THREAD

**Issue by maheshdatt** [xetorthio/jedis#1750]
### Expected behavior

When a redis server is restarted (ungraceful - by pulling a plug) - I expect the clients to reconnect seemlessly.

### Actual behavior

We noticed during an ungraceful redis server restart, our jedis client was hanging indefinitely since the underlying socket timeout was infinite.

### Steps to reproduce:

BRPOP code in the jedis 2.9.0 client looks like this: 

```
public List<String> brpop(final String... args) {
    checkIsInMultiOrPipeline();
    client.brpop(args);
    client.setTimeoutInfinite();
    try {
      return client.getMultiBulkReply();
    } finally {
      client.rollbackTimeout();
    }
}
```
The part thats confusing is the setting the client socket timeout to infinite. Why do we do it is there a good reason for it? Is there a way to override this to a finite value? 

The reason I m asking is because in our production env, we encountered an issue with redis recently where we were doing a brpop on a redis server. This physical machine thats running redis got hard rebooted and the brpop on the client blocked for 2 hours (the OS socket timeout). 

We would like to know if there is a setting that can set it to lower number?

### Redis / Jedis Configuration
redis nodes are running in non-cluster mode. Client nodes pop data using jedis connection pool.

#### Jedis version:
2.9.0
#### Redis version:
2.8.24

**Comment by maheshdatt** [comment]
There seem to be another request like this a while ago which was closed without resolution:
https://github.com/xetorthio/jedis/issues/426

**Comment by sazzad16** [comment]
@maheshdatt First of all, it is not clear which overloaded method you're actually calling. 

However, could you check what happens in more recent versions of Redis? I tried to reproduce your issue and I always get `redis.clients.jedis.exceptions.JedisConnectionException: Unexpected end of stream.` after I kill redis server in all overloaded versions except the deprecated one (`List<String> brpop(String arg)`).

**Comment by maheshdatt** [comment]
@sazzad16 - appreciate your response. 

we are using the method (I also updated the above code snippet. Somehow it didn't show earlier due to formatting issue.)

```
public List<String> brpop(final int timeout, final String... keys)
```

This is my sample client side code:

```
        for(;;) {
            try(Jedis jedis = getJedisPool().getResource())  {
                System.out.println("Requesting brpop");
                System.out.println("Returned from brpop: "+jedis.brpop(0,"myqueue"));
            } catch (Exception ex) {
                ex.printStackTrace();
            }
        }
```
To reproduce this the redis server need to be hard rebooted. I was able to easily reproduce this issue while running redis on vagrant VM. Then I "reset" the VM via Oracle VirtualBox VM manager.
That makes the jedis brpop to hang indefinitely. 

I doubt changing redis version helps since it seems to be an issue with jedis client overriding the socket timeout to infinite. I am not sure why its overridden to infinite inspite of providing an option to specify socket timeout.

**Comment by sazzad16** [comment]
@maheshdatt I tried your example. Got one `redis.clients.jedis.exceptions.JedisConnectionException: Unexpected end of stream.` right after killing redis and several `redis.clients.jedis.exceptions.JedisConnectionException: Could not get a resource from the pool` until redis is up again.

Just so you know, you're using `0` as timeout which is infinite. Maybe your list `myqueue` is empty and because of that redis is waiting for indefinite amount of time?

**Comment by maheshdatt** [comment]
@sazzad16 - I understand that I m providing 0 for timeout - which is infinite blocking. But the client should still be able to determine that the remote end has exited and close the connection.

So in my case, after the redis server is hard rebooted, even when I do ```lpush``` to put data onto ```myqueue``` I still was not able to dequeue via brpop. 

The only solution that works for me is to implement a background pinger thread that independently sends ```jedis.ping()``` and kills any hanged ```brpop``` connections. This will make it acquire a new connection and then retrieve the data.

I am pretty sure in your case redis did shutdown gracefully and closed the socket which resulted in client side connection exceptions. If you can do an abrupt shutdown like pulling the network cable/powering down the machine, you will be able to reproduce my scenario.

This is my sample thread program that pings redis independently and closes open connections in case they are broken.

```
        public void run() {
            for(;;) {
                try (Jedis jedis1 = getJedisPool().getResource()) {
                    jedis1.ping();
                } catch (Exception e) {
                    try {
                        log.info("Closing jedis connection since it may be broken");
                        if (jedis.isConnected()) {
                            jedis.close();
                        }
                        break;
                    }catch (Exception ex) {
                        log.info("Error closing jedis connection : {}", ex.getMessage());
                    }
                }
            }
        }
```


Thanks for your help!

**Comment by sazzad16** [comment]
@maheshdatt Got it!

**Comment by maheshdatt** [comment]
@sazzad16 - were you able to reproduce it ? 

Is there a workaround to this problem? Thanks.

**Comment by sazzad16** [comment]
@maheshdatt How about setting a time limit on your method?

**Comment by sazzad16** [comment]
@maheshdatt How about setting a time limit on your method? Don't forget to invalidate the Jedis object if it fails.

**Comment by maheshdatt** [comment]
@sazzad16  What exactly do you mean by "time limit on your method" ? 

Are you referring to specifying the timeout inside brpop? That wouldn't catch this scenario since the socket time is overridden to infinite inside the jedis client library.

**Comment by sazzad16** [comment]
@maheshdatt I meant setting timeout on **your** implementation, **outside** brpop. Something like https://stackoverflow.com/questions/5715235/java-set-timeout-on-a-certain-block-of-code

**Comment by maheshdatt** [comment]
We tried that approach - but felt hacky. At this point we are exploring the option of forking jedis and making the blocking honor the application timeout if they are provided, instead of setting the socket timeout to infinite.

**Comment by sazzad16** [comment]
Closed as dupe of #426

**Comment by ZZ920703114** [comment]
I had this problem too, my logs print out with ‘’Unexpected end of stream‘’.and ‘’Could not get a resource from the pool‘’，
I set the timeout of brpop to 30s, redis to 31s, and mysql database connection pool to 300s. For now, this error is not printed on the log, but I don't know if this error will be reported in the future.
